scaletest/workspacetraffic/conn.go

```diff
@@ -2,122 +2,260 @@ package workspacetraffic
+	//
+	// Since we can't control fragmentation/buffer sizes, we keep it simple and
+	// match the conservative payload size used by agent/reconnectingpty (1024).
+	rptyJSONMaxDataSize = 1024
+)
+
+func connectRPTY(ctx context.Context, client *codersdk.Client, agentID, reconnect uuid.UUID, cmd string) (*countReadWriteCloser, error) {
+	width, height := 80, 25
 	conn, err := client.WorkspaceAgentReconnectingPTY(ctx, codersdk.WorkspaceAgentReconnectingPTYOpts{
 		AgentID:   agentID,
 		Reconnect: reconnect,
-		Height:    25,
-		Width:     80,
-		Command:   "sh",
+		Width:     uint16(width),
+		Height:    uint16(height),
+		Command:   cmd,
 	})
 	if err != nil {
 		return nil, xerrors.Errorf("connect pty: %w", err)
 	}
 
 	// Wrap the conn in a countReadWriteCloser so we can monitor bytes sent/rcvd.
-	crw := countReadWriteCloser{ctx: ctx, rwc: conn}
+	crw := countReadWriteCloser{rwc: newPTYConn(conn)}
 	return &crw, nil
 }
 
-func connectSSH(ctx context.Context, client *codersdk.Client, agentID uuid.UUID) (*countReadWriteCloser, error) {
+type rptyConn struct {
+	conn io.ReadWriteCloser
+	wenc *json.Encoder
+
+	mu       sync.Mutex
+	closed   bool
+	readErr  chan error
+	readOnce sync.Once
+}
+
+func newPTYConn(conn io.ReadWriteCloser) *rptyConn {
+	rc := &rptyConn{
+		conn:    conn,
+		wenc:    json.NewEncoder(conn),
+		readErr: make(chan error, 1),
+	}
+	return rc
+}
+
+func (c *rptyConn) Read(p []byte) (int, error) {
+	n, err := c.conn.Read(p)
+	if err != nil {
+		c.readOnce.Do(func() {
+			c.readErr <- err
+			close(c.readErr)
+		})
+		return n, err
+	}
+	return n, nil
+}
+
+func (c *rptyConn) Write(p []byte) (int, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// Early exit in case we're closing, this is to let call write Ctrl+C
+	// without a flood of other writes.
+	if c.closed {
+		return 0, io.EOF
+	}
+
+	return c.writeNoLock(p)
+}
+
+func (c *rptyConn) writeNoLock(p []byte) (n int, err error) {
+	// If we try to send more than the max payload size, the server will close the connection.
+	for len(p) > 0 {
+		pp := p
+		if len(pp) > rptyJSONMaxDataSize {
+			pp = p[:rptyJSONMaxDataSize]
+		}
+		p = p[len(pp):]
+		req := codersdk.ReconnectingPTYRequest{Data: string(pp)}
+		if err := c.wenc.Encode(req); err != nil {
+			return n, xerrors.Errorf("encode pty request: %w", err)
+		}
+		n += len(pp)
+	}
+	return n, nil
+}
+
+func (c *rptyConn) Close() (err error) {
+	c.mu.Lock()
+	if c.closed {
+		c.mu.Unlock()
+		return nil
+	}
+	c.closed = true
+	c.mu.Unlock()
+
+	defer c.conn.Close()
+
+	// Send Ctrl+C to interrupt the command.
+	_, err = c.writeNoLock([]byte("\u0003"))
+	if err != nil {
+		return xerrors.Errorf("write ctrl+c: %w", err)
+	}
+	select {
+	case <-time.After(connCloseTimeout):
+		return xerrors.Errorf("timeout waiting for read to finish")
+	case err = <-c.readErr:
+		if errors.Is(err, io.EOF) {
+			return nil
+		}
+		return err
+	}
```

Comment:
I don't think we need to, I opted to use the `closed` boolean to allow `Write` to exit as soon as we hit `Close()`. If we held the lock throughout, we wouldn't need the boolean but it might give the wrong idea up the stack about why the write routine didn't exit in time 🤔.

Reply:
I reorganized the struct to clarify why `mu` is present.